DEV-66 DEV-245 DEV-261 DEV-266 Rubric improvements

olmokramer opened the pull request:
This improves rubrics in CodeGrade in multiple ways:

* You can now reorder rubric categories. The UI of the rubric editor has been
  rewritten to get rid of the messy tabs at the top. Instead the categories are
  listed vertically on the left side of the rubric editor.

* You can now use markdown in rubric descriptions. A new component,
  PreviewableMarkdownEditor, was introduced that is a regular text area with
  controls to preview the written markdown.

* Markdown rendering has improved. Code blocks now have a distincive background
  color, and the text size in headings is scaled relative to the rest of the
  document rather than to the page root.

**TODO**

* [x] Fix tests
* [x] Open/close all button
* [x] Add possibility to change between old and new rubric editor after having chosen one already.

DEV-66 #done
DEV-245 #done
DEV-261 #done
DEV-266 #done

libre-man commented on migrations/versions/cd3c64fb78d7_.py:
```diff
@@ -0,0 +1,62 @@
+        for i, row in enumerate(sorted_rows):
+            bind.execute(
+                text('''
+                UPDATE "RubricRow" SET position = {}
```
We should sql parameters instead of formatting the string.

libre-man commented on psef_test/test_assignments.py:
```diff
@@ -431,6 +432,54 @@ def test_update_rubric_row(
 
+def test_reorder_rubric_rows(
+    logged_in, test_client, teacher_user, assignment
+):
```
Could you maybe add comments or use the `describe` in this test?

libre-man commented on src/models/rubric.ts:
```diff
@@ -542,6 +561,36 @@ export class Rubric<T extends number | undefined | null> {
+
+    moveRowRight(idx: number) {
+        if (idx >= this.rows.length - 1) {
+            console.log('nothing changed');
```
Please remove

olmokramer replied:
Ah we should remove the entire method as it isn't used anymore. Also `moveRowLeft`.

libre-man commented on src/components/PreviewableMarkdownEditor.tsx:
```diff
@@ -0,0 +1,74 @@
+// @ts-ignore
+import InnerMarkdownViewer from './InnerMarkdownViewer';
+
+const PreviewableMarkdownEditor = tsx.component({
```
Not really important for now, but you can also type the possible emits from these tsx components.

olmokramer replied:
Oh that's cool! I've typed the events emitted by this component.

libre-man commented on src/components/RubricEditor.vue:
```diff
@@ -74,128 +74,184 @@
+                                variant="danger"
+                                class="delete-category"
+                                :wait-at-least="0"
+                                :submit="() => {}"
```
Shouldn't this button do something?

libre-man commented on migrations/versions/cd3c64fb78d7_.py:
```diff
@@ -0,0 +1,66 @@
+"""Add 'position' column to RubricRow table and order all existing
```
It seems like there are no tests for the migration, which I think is really important here.

libre-man commented on src/components/RubricEditor.vue:
```diff
@@ -74,128 +74,205 @@
+    <template v-if="rubricRows.length === 0">
+        <h4 v-if="editable"
+            slot="empty"
+            class="no-categories flex-grow-1 mb-3 p-5 border rounded text-center text-muted">
```
I think it makes more sense to automatically add a first category when a user clicks on "Create a new rubric" (like we want to do for AutoTest steps too).

Besides that I think the text is a bit large here, but not really important.

olmokramer replied:
Once we split the <kbd>+ Category</kbd> button into a button per category type (which I think is an excellent idea!), I don't think we would gain a lot from creating a new category by default, because the user would still have to select the type of the category. In fact, I think it might cause confusion as there are two sets of very distinctly looking buttons that do the exact same thing.

About the large text, that hasn't changed. I made the text here a bit larger because otherwise the page would be so empty without categories.

libre-man commented on src/components/RubricEditor.vue:
```diff
@@ -74,128 +74,205 @@
+            </slick-item>
+        </transition-group>
+
+        <b-button-toolbar v-if="editable" class="justify-content-center mb-3">
```
I think aligning this button to the right is a bit more consistent with the rest of the app. Not really important though.

olmokramer replied:
I center-aligned it to make it look similar to the AutoTest suite. And I'm really in favour of splitting this button, so then it will look a bit more similar to it.

libre-man commented on src/components/RubricEditor.vue:
```diff
@@ -74,128 +74,205 @@
-                <template v-if="row.header">
-                    {{ row.header }}
-                </template>
+    <template v-if="rubricRows.length === 0">
```
I would really like to see "expand all" and "collapse all" buttons.

olmokramer replied:
Ah right, forgot to add those.

olmokramer replied:
Done!

libre-man reviewed (commented):
## A few things about the editor in non editable mode:
- [x] If there is only one item I would simply expand it (this is a common case with AutoTest).
- [x] I would move the note about the amount of points needed to get a full mark to the top of the items. Because currently with many items it could be place under the fold quite easily.
- [ ] I think it makes sense to also state how many items there are in the category for discrete rubric items. Not sure how we would do this though...

## A few thinks about the editor when editing:
- [x] I'm not entirely sure why the submit button now uses a confirm in a modal (we talked about it be I forgot), but I find it very annoying as it the message isn't smart about new possible mistakes so when you have a rubric in which we think there is a mistake we always show this modal (and modals are simply very annoying). Oh it is because of <kbd>C-RET</kbd>, is it possible to only display the confirm the modal in this case?
- [x] I think that the min height for the description (items and categories) could be a bit less. It used to not really matter as we had enough vertical space anyway, but now it does matter a bit more and with the rubrics I checked the description is usually quite small.
- [x] The collapse and uncollapse animation is very smooth, nicely done! The animation when creating a new item is also very nice.
- [ ] About the add category button, I think it would make more sense to maybe have two buttons, one for a discrete category and one for continuous, just like we have with AutoTest steps. In this case we should keep the buttons centered.

libre-man commented on src/models/rubric.ts:
```diff
@@ -142,12 +167,19 @@ export class RubricRow<T extends number | undefined | null> {
             throw new Error('You cannot make a base row with a non empty type.');
         }
+
+        this.trackingId = getUniqueId();
```
I'm not really a fan of how this is done, as we now keep the same `trackingId`  when updating by the fact that it is present in the argument (as an extra key which is valid). I would prefer that we only copy the requested keys from the argument `row` (using `utils.pickKeys`) and simply pass `trackingId` as a second (optional) argument to the constructor.

olmokramer replied:
Yeah I agree. I've changed it as per your suggestion.

olmokramer commented on src/components/PreviewableMarkdownEditor.tsx:
```diff
@@ -0,0 +1,90 @@
+const PreviewableMarkdownEditor = tsx.componentFactoryOf<Events, {}>().create({
+    props: {
+        value: p(String).required,
+        rows: p(Number).default(0),
```
I chose a `0` because that makes the browser use its own default value, but yeah that's not very clear so I've changed the type of the `rows` prop to accept `number | undefined` so that it doesn't set the `rows` attr on the textarea at all and to make this more explicit. I don't think a default of `1` makes a lot of sense, because if we want something to be a single line input we shouldn't use a `<textarea>` but rather a `<input>`.

olmokramer commented on src/components/PreviewableMarkdownEditor.tsx:
```diff
@@ -0,0 +1,90 @@
+        },
+    },
+
+    render(h: CreateElement) {
```
Yeah I agree this is so much better!

olmokramer commented on src/components/RubricEditor.vue:
```diff
@@ -74,128 +74,205 @@
-                        rubric to achieve the maximum grade. By default students
-                        must achieve the top item in each discrete category and
-                        100% in each continuous category.<br>
+                        <p class="mb-2">
```
Because by default `<p>`s have a bottom margin of `1rem`, which is a bit much inside a popover. We have this in other places as well... But maybe we should set a global style for this, something like

```scss
.popover {
    p, ul, pre /* ... */ {
        margin-bottom: 0.5rem;
        /* or once we have switched to sass */
        @extend .mb-2;
    }
}
```

libre-man replied:
Yeah that would be a good idea, but you don't have to do this in this PR.

olmokramer commented on src/components/RubricEditor.vue:
```diff
@@ -219,172 +296,183 @@
+                                  confirm-in-modal
+                                  :submit="submit"
+                                  @after-success="afterSubmit">
+                    <div slot="confirm" class="text-justify">
```
Yeah there's a lot of duplication there, though the messages we show in the confirm are different from the messages we show in the error popover. Some simple component that could render below template would help a lot I think, and for that tsx will help a lot.

```vue
<b>{{ header }}</b>

<p v-for="p in paragraphs">{{ p }}</p>

<ul>
  <li v-for="msg in messages">{{ msg }}</li>
</ul>
```

olmokramer commented:
I've addressed some of your comments in https://github.com/CodeGra-de/CodeGra.de/pull/1471/commits/7e21a11a4f8ce1ddaee5f502fc70b7f2bff0ed9b. See the commit message for the things I've done already.

olmokramer commented:
PS I'm also changing your comment to have checkboxes in the list so I can keep track of those points as well

libre-man commented:
> PS I'm also changing your comment to have checkboxes in the list so I can keep track of those points as well

Good idea!

BTW: Really liked the workflow when adding new rubrics, it is so much easier to work on two categories at the same time now!

olmokramer commented:
> ## A few things about the editor in non editable mode:
> 
> * If there is only one item I would simply expand it (this is a common case with AutoTest).

Yeah that's a great idea. Now that I've implemented it I also think this single category should not be collapsible.

> * I would move the note about the amount of points needed to get a full mark to the top of the items. Because currently with many items it could be place under the fold quite easily.

Yeah I agree that it should be easily visible. But it is kind of annoying to have it above the input, because as soon as you start typing in the input it jumps down because the rather large message is now suddenly visible. Still I've placed the message at the top for now, but I hope we can think of something better in the future.

> * I think it makes sense to also state how many items there are in the category for discrete rubric items. Not sure how we would do this though...

What about the following idea: instead of showing the range of points you can achieve, for discrete categories we show `(x / y / z pts.)` for a category where you can get `x`, `y`, or `z` points. Then it's also immediately clear if a category is discrete or continuous and in a useful way because it is actually telling you how you can score points in that category. Only issue I see with this is that it may get a bit long for rows with a lot of categories. Maybe in those cases `x / y / ... / z` or `x / y / ... N ... / z` (where `N` is the amount of categories not in `{ x, y, z }`) would be enough, especially if the number of points grows linearly (though I don't know if that's the norm).

> ## A few thinks about the editor when editing:
> 
> * I'm not entirely sure why the submit button now uses a confirm in a modal (we talked about it be I forgot), but I find it very annoying as it the message isn't smart about new possible mistakes so when you have a rubric in which we think there is a mistake we always show this modal (and modals are simply very annoying). Oh it is because of C-RET, is it possible to only display the confirm the modal in this case?

Yeah that was because of <kbd>Ctrl+Enter</kbd>. I've implemented your suggestion of only using the modal with <kbd>Ctrl+Enter</kbd>.

> * I think that the min height for the description (items and categories) could be a bit less. It used to not really matter as we had enough vertical space anyway, but now it does matter a bit more and with the rubrics I checked the description is usually quite small.

Yeah the descriptions (especially the category descriptions) aren't usually that long, so it's fine to decrease them a bit. I've changed the category descriptions from 5 to 3 lines and the item descriptions from 8 to 5 lines.

> * The collapse and uncollapse animation is very smooth, nicely done! The animation when creating a new item is also very nice.

Thanks! Yeah they turned out pretty nicely! Although all I did was copy the code from `AutoTestSuite.vue` and then basically nothing...

> * About the add category button, I think it would make more sense to maybe have two buttons, one for a discrete category and one for continuous, just like we have with AutoTest steps. In this case we should keep the buttons centered.

Yeah that's also a great idea.

libre-man commented:
> What about the following idea: instead of showing the range of points you can achieve, for discrete categories we show (x / y / z pts.) for a category where you can get x, y, or z points. Then it's also immediately clear if a category is discrete or continuous and in a useful way because it is actually telling you how you can score points in that category. Only issue I see with this is that it may get a bit long for rows with a lot of categories. Maybe in those cases x / y / ... / z or x / y / ... N ... / z (where N is the amount of categories not in { x, y, z }) would be enough, especially if the number of points grows linearly (though I don't know if that's the norm).

Yeah that is a great idea! I wouldn't worry too much about many items, they are always displayed on a single line in the viewer, so if we can't show them in these headers we will never be able to actually display them on the submission page. If we ever fix it (or/and if it becomes a problem) your propose solution seems great.

libre-man commented:
Another thing I noticed: in the updated rubric viewer (not a great fan of the reduced padding, but I don't really care that much) it seems like there is a border below the active tab, which is not something the other tabs have, and I think it would look better if we don't do that.

olmokramer commented:
> Another thing I noticed: in the updated rubric viewer (not a great fan of the reduced padding, but I don't really care that much) it seems like there is a border below the active tab, which is not something the other tabs have, and I think it would look better if we don't do that.

Yeah I think that was because of the semi-transparent background of the tabs that overlapped the border of the description block... I've added an opaque `@card-header-background` and `@dark-card-header-background` now that is the same as bootstrap's card header background in the light theme, and slightly lighter than card headers in the dark theme, because

libre-man commented:
Almost there! Just a few minor things:

- [ ] Really like the two buttons to add the two types of categories. I'm a bit scared though that users won't know where to click to add a new category. We do explain it when there are no categories at all, but as the text is muted I'm not that hopeful that people will actually read the text. Besides that, if users revisit a rubric somebody else (or they a week ago) created the text won't be there and the two buttons don't really "feel" like buttons to add new stuff. Don't really know how to solve this though, maybe replace the icons with `fa-plus` icon and the text with `Add $type category`?
- [ ] I understand that we need some sort of place to put the expand/collapse all button, but I think that the "Categories" header is a bit weird, especially as it is the only header in the card/on the page. Maybe we can place the buttons all the way on the right side? Don't know for sure, lets discuss it for a bit IRL.
- [ ] I think your idea about not making rubrics with a single category collapsible is great. Do we still want to do it in this PR or do it in another one? Furthermore, it seams like it is not yet expanding on the rubric editor, is that intentional, and if yes, why?
- [ ] I think we should disable the submit button when all items are deleted, instead of simply showing an error when it is clicked.

libre-man commented on src/components/PreferredUI.tsx:
```diff
@@ -0,0 +1,110 @@
+
+    computed: {
+        userId() {
+            return store.getters['user/id'];
```
These are all probably typed as `any`. I think it would make sense to narrow the type a bit down in the annotations.

olmokramer replied:
Yes, good catch

libre-man commented on src/components/PreferredUI.tsx:
```diff
@@ -0,0 +1,110 @@
+
+        renderError(h: CreateElement) {
+            return this.error.mapOrDefault(
+                e => <comp.CgError error={e} /> as VNode,
```
Why are these casts necessary?

olmokramer replied:
Very strange, and I haven't been able to figure out why this happens yet, but typescript infers `<comp.CgError error={e} />` to be of type `Element` rather than `VNode`, even though `CreateElement` is the type of a function that returns a `VNode`.

olmokramer replied:
Oh, and If I remove the cast from the default case typescript infers the type of the entire `this.error.mapOrDefault(...)` expression to be `EmptyVNode`, and neither an `Element` nor a `VNode` is assignable to type `EmptyVNode`.

Rewritten as below typescript accepts the function without casts.

```tsx
return utils.ifOrEmpty(
    this.error.isJust(),
    () =>  <comp.CgError error={this.error.unsafeCoerce()} />,
);
```

libre-man commented on src/components/PreferredUI.tsx:
```diff
@@ -93,17 +91,73 @@ export default tsx.component({
         },
 
         updatePreference(value: boolean) {
+            this.updating = true;
```
I think that taking an optimistic UI approach makes a bit more sense here. So not displaying the loader and simply updating this value in the background. If the request goes wrong (low chance) nothing important is lost (we could even display a toast in this case) but you don't have to wait for a request when switching. Not that important, and we can do it later.

olmokramer replied:
Yeah that's what I tried first, but it didn't look nice or feel smooth at all. But now that I've tried it again it's much better, so something has changed although I don't know what... Anyway, I've removed the `updating` property and its associated loader.

libre-man commented on src/components/PreferredUI.tsx:
```diff
@@ -53,33 +54,30 @@ export default tsx.component({
-            const value = this.prefValue.orDefault(this.defaultValue);
-            return value ? this.$slots.ifTrue : this.$slots.ifFalse;
+        renderError(h: CreateElement): VNode {
+            return utils.ifOrEmpty(
```
Maybe we should make a wrapper function for this: `justOrEmpty` that utilizes the `caseOf` method of the `Maybe`. Not important though.

olmokramer replied:
Yeah I didn't do that yet because it would've been useful only once in this component, unless I rewrite `this.allPrefs.isJust() && this.prefValue.isJust()` in the `renderToggle` method as `this.allPrefs.map(this.prefValue).join()` before passing it in to `ifJustOrEmpty`, but I don't think that makes it more readable.

olmokramer replied:
I've added the function anyway, as it's very likely to be useful in the future as well.

libre-man commented:
- I also get the rubric switcher when there is no rubric on the submissions page.
- In my opinion the danger button (to switch to the old version) in the alert is really ugly. I'm not so sure it should be a danger button anyway, as you don't do anything that can't be reversed, or anything really dangerous at all. I also think we could reduce the text quite a bit. Maybe @yourivoet or @DevinHillenius can give some input about it?
- I think it is a bit confusing that the select to switch versions is below, while the alert is above. I understand why (it is quite large and takes quite a bit of space), but I'm not really a fan of the design of the switcher anyway as I think it is quite prominent. Also I don't think a `select` is the most obvious choice here, I discussed it a bit with @yourivoet and he suggested simply adding a small button/link that switches back to the other version. So maybe something like this: 
![image](https://user-images.githubusercontent.com/12597247/95219162-59359d80-07f5-11eb-827a-537b1bee1b75.png)

libre-man commented:
When the rubric editor is editing mode the switcher now doesn't show up if the assignment doesn't have a rubric. It also doesn't show up after I create a rubric, it simply switches to the old interface. It starts to show up when I first submit the new rubric.

libre-man commented on src/components/RubricEditor/index.tsx:
```diff
@@ -0,0 +1,57 @@
+    name: 'rubric-editor-wrapper',
+
+    props: {
+        assignment: p.ofType<models.Assignment>().required,
```
As this is an actual model I think we can use `p(models.Assignment)` here which has the advantage of being checked by `vue` in dev mode too.

olmokramer replied:
No we can't, that produces this error

```
[tsl] ERROR in /home/olmo/.local/src/CodeGra.de/src/components/RubricEditor/index.tsx(20,23)
      TS2345: Argument of type 'typeof Assignment' is not assignable to parameter of type 'Prop<Assignment>'.
  Type 'typeof Assignment' is not assignable to type 'new (...args: string[]) => Function'.
    Cannot assign a 'private' constructor type to a 'public' constructor type.
```

and it looks like there's no way around that, because the constructors passed to `p` must be of type [`Prop<T>`](https://github.com/vuejs/vue/blob/668e1e637461ff630803e85bf99158415d276d4c/types/options.d.ts#L147) while our models aren't.

libre-man replied:
Ah that is annoying, well then we'll simply use this. I don't think it makes sense to add a validator here.

olmokramer commented:
> When the rubric editor is editing mode the switcher now doesn't show up if the assignment doesn't have a rubric. It also doesn't show up after I create a rubric, it simply switches to the old interface. It starts to show up when I first submit the new rubric.

Ah yes good catch. I've fixed the check.
